core/src/main/scala/cats/Foldable.scala

```diff
@@ -433,4 +448,26 @@ object Foldable {
     }
     M.tailRecM(z)(go)
   }
+
+
+  private sealed abstract class Source[+A] {
+    def uncons: Option[(A, Source[A])]
+  }
+
+  private object Source {
+    val Empty: Source[Nothing] = new Source[Nothing] {
+      def uncons = None
+    }
+
+    val NowEmpty: Eval[Source[Nothing]] = Now(Empty)
+
+    def cons[A](a: A, src: Eval[Source[A]]): Source[A] = new Source[A] {
+      def uncons = Some((a, src.value))
```

Comment:
The presence of `Eval` in this case actually just obstructs the understanding, IMO. The `src` parameter of `cons` could just be by-name (the evaluation of `.value` doesn't need any trampolining built into `Eval`). However, since `foldRight` already uses `Eval`, I didn't want to mix both by-name and `Eval`.

A clearer definition of `Source` would be

```scala
type Source[A] = () => Option[(A, Source[A])]
```

except that recursive type aliases are not allowed. I guess I could make it an `AnyVal`, though.